### PR TITLE
qcontainer: Don't set bus for devices which have no 'bus' option

### PR DESCRIPTION
**Pull request by huangyum**
ID: 1597464
Signed-off-by: Yumei Huang <yuhuang@redhat.com>

### CODE REVIEW

**Comment by huangyum** [comment]
Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.hotplug.backend_ram.policy_default.two: PASS (206.70 s)

**Comment by vivianQizhu** [comment]
ACK

**Comment by jingzhao84** [comment]
ack
 Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.hotplug.backend_ram.policy_default.two: PASS

**Comment by jingzhao84** [comment]
@MiriamDeng  could you help to check ppc part?

Thanks
Jing

**Comment by MiriamDeng** [comment]
ACK.
 The patch works well on ppc.
 (1/1) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.ppc64le.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.hotplug.backend_ram.policy_default.two: PASS (194.78 s)

**Comment by jingzhao84** [comment]
@luckyh  could you help to review this pr? thanks

**Comment by huangyum** [comment]
@luckyh I can provide another two ways to solve the problem, but not sure if any one of them is fine. Please help review. 
1. Overwrite `set_param` for Dimm class, so that `prepare_hotplug` won't set bus for Dimm.
2. Overwrite `hotplug_qmp` for Dimm class, pop 'bus' out when generate hotplug cmd. 
Any better idea is welcome too, thanks！

**Comment by huangyum** [comment]
I just came up with a third solution to check if the device has 'bus' option before set bus for it. Please help review, thanks! @luckyh

**Comment by huangyum** [comment]
Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.hotplug.backend_ram.policy_default.two: PASS (207.36 s)

File: virttest/qemu_devices/qcontainer.py [hunk 1]
```diff
@@ -597,7 +597,8 @@ def simple_hotplug(self, device, monitor, bus=None):
         """
         self.set_dirty()
 
-        if isinstance(device, qdevices.QDevice):
+        if isinstance(device, qdevices.QDevice) and \
+                "bus" in self.execute_qemu("-device %s,?" % device["driver"]):
```
Comment (vivianQizhu) [inline virttest/qemu_devices/qcontainer.py]: That's not safe, actually most QDevice has `bus` option does not include `bus` in its `-device %s,?`, e.g. `nec-usb-xhci`, `virtio-pci` has `bus` but not exactly option `bus`, it just luckily has an option `virtio-pci-bus-master-bug-migration=bool (on/off)`

Comment (vivianQizhu) [reply virttest/qemu_devices/qcontainer.py]: It looks like `/usr/libexec/qemu-kvm -device ?` will give exactly the device has bus or not.

**Comment by luckyh** [comment]
@huangyum the problem is at here:
```python
if isinstance(device, qdevices.QDevice):
    if bus is None:
        bus = self.get_buses({'aobject': 'pci.0'})[0]  # <--
```
Not all devices would attach to `pci.0`, for example, powerpc spapr devices would attach to spapr bus, and yes dimm device is one of them, so that means we should not hardcode the default bus at above.

A feasible solution is to configure a default bus for each type of devices (`-device [driver],...` in qemu command line). For those devices who don't have an obvious parent bus, we can assign a pseudo bus for them.

**Comment by vivianQizhu** [comment]
@luckyh Actually, currently code works fine for all QDevice that has a `bus` option, even not pci bus device. line604 - line609 will assign the parent bus of device instead of pci.0 to `bus`, it can resolve any `QDevice` bus assignment if the `QDevice` has proper `parent_bus` been assigned(like scsi-hd will be assigned to virtio-scsi-pci instead of pci.0, usb device will be assigned to usb controller). Same, spapr device will be assigned to spapr bus if the device has spapr bus as its parent_bus in `QDevice` object.

The real problem is Dimm has no `bus`.

**Comment by vivianQizhu** [comment]
@luckyh And that's actually does not matter what qemu command line is configured, that's all about how to configure `device_add` command(QDevice object hotplug = device_add). 
Besides, none QDevice type device(like QStringDevice or QObject) will not be assgined `bus`, currently only class Dimm() inherited from QDevice has this issue(no `bus` option)

File: virttest/qemu_devices/qcontainer.py (hunk 1 above)
Comment (huangyum) [reply virttest/qemu_devices/qcontainer.py]: @vivianQizhu  I thought `-device ?` would show all the options it accepts, but apparently I was wrong.  I checked everywhere but didn't find a way to list all the available options.  Seems this approach is not workable.

Comment (vivianQizhu) [reply virttest/qemu_devices/qcontainer.py]: @huangyum `-device ?` actually shows the bus of device, e.g.:
```
Input devices:
name "ccid-card-passthru", bus ccid-bus, desc "passthrough smartcard"
name "isa-serial", bus ISA
name "pci-serial", bus PCI
```
```
Uncategorized devices:
name "edu", bus PCI
name "intel-iommu", bus System
name "nvdimm", desc "DIMM memory module"
name "pc-dimm", desc "DIMM memory module"
name "virtio-gpu-device", bus virtio-bus
name "vmgenid"
```
Exactly shows the device has bus or not, and what is the bus type of the device.

**Comment by huangyum** [comment]
@luckyh I agree with @vivianQizhu . Dimm device does not accept "bus" option, that's the real issue.  Maybe we should just take it as an exception ?

**Comment by vivianQizhu** [comment]
@luckyh I have tried to assign only pci device the pci bus https://github.com/avocado-framework/avocado-vt/pull/1653 , what do you think of it?

**Pull request closed without merging**